File: trash_from_senders.py
#!/usr/bin/env python3
"""
指定送信元ドメインのメールを Gmail ゴミ箱へ一括移動する。

既定は dry-run（件数・ドメイン別内訳のみ表示）。
実際の移動は --exec 指定時のみ。完全削除は行わない。

認証:
  credentials/credentials.json … OAuth クライアント（Desktop 推奨）
  credentials/token.json … 初回認可後に自動保存

スコープ: https://www.googleapis.com/auth/gmail.modify
"""

from __future__ import annotations

import argparse
import sys
from collections import Counter
from pathlib import Path

ROOT = Path(__file__).resolve().parent
CRED_DIR = ROOT / "credentials"
CREDENTIALS_FILE = CRED_DIR / "credentials.json"
TOKEN_FILE = CRED_DIR / "token.json"

SCOPES = ["https://www.googleapis.com/auth/gmail.modify"]

DEFAULT_QUERY = "from:nikkeibp.co.jp OR from:media-radar.jp"
DOMAIN_KEYS = ("nikkeibp.co.jp", "media-radar.jp")
BATCH_SIZE = 1000
LIST_PAGE_SIZE = 500


def get_gmail_service():
    try:
        from google.auth.transport.requests import Request
        from google.oauth2.credentials import Credentials
        from google_auth_oauthlib.flow import InstalledAppFlow
        from googleapiclient.discovery import build
    except ImportError as e:
        print(
            "必要なパッケージがありません。次を実行してください:\n"
            "  pip3 install google-api-python-client google-auth-oauthlib",
            file=sys.stderr,
        )
        raise SystemExit(1) from e

    if not CREDENTIALS_FILE.exists():
        print(
            f"OAuth クライアントが見つかりません: {CREDENTIALS_FILE}\n"
            "Google Cloud Console で Desktop アプリの OAuth クライアントを作成し、\n"
            "JSON を credentials/credentials.json として配置してください。\n"
            f"必要なスコープ: {SCOPES[0]}",
            file=sys.stderr,
        )
        raise SystemExit(2)

    creds = None
    if TOKEN_FILE.exists():
        creds = Credentials.from_authorized_user_file(str(TOKEN_FILE), SCOPES)

    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            granted = set(creds.scopes or [])
            if set(SCOPES).issubset(granted):
                try:
                    creds.refresh(Request())
                except Exception:
                    creds = None
            else:
                creds = None
        else:
            creds = None

    if not creds or not creds.valid:
        flow = InstalledAppFlow.from_client_secrets_file(
            str(CREDENTIALS_FILE), SCOPES
        )
        creds = flow.run_local_server(port=0)
        print("認可完了。token.json を更新しました。")

    CRED_DIR.mkdir(parents=True, exist_ok=True)
    TOKEN_FILE.write_text(creds.to_json(), encoding="utf-8")

    return build("gmail", "v1", credentials=creds)


def list_all_message_ids(service, query: str) -> list[str]:
    ids: list[str] = []
    page_token = None
    while True:
        resp = (
            service.users()
            .messages()
            .list(
                userId="me",
                q=query,
                pageToken=page_token,
                maxResults=LIST_PAGE_SIZE,
            )
            .execute()
        )
        for m in resp.get("messages") or []:
            mid = m.get("id")
            if mid:
                ids.append(mid)
        page_token = resp.get("nextPageToken")
        if not page_token:
            break
    return ids


def classify_by_domain(service, message_ids: list[str]) -> Counter:
    """From ヘッダを軽く取得してドメイン別内訳を出す（dry-run / 報告用）。"""
    counts: Counter = Counter()
    for mid in message_ids:
        meta = (
            service.users()
            .messages()
            .get(
                userId="me",
                id=mid,
                format="metadata",
                metadataHeaders=["From"],
            )
            .execute()
        )
        headers = (meta.get("payload") or {}).get("headers") or []
        from_val = ""
        for h in headers:
            if h.get("name", "").lower() == "from":
                from_val = (h.get("value") or "").lower()
                break
        matched = False
        for domain in DOMAIN_KEYS:
            if domain in from_val:
                counts[domain] += 1
                matched = True
                break
        if not matched:
            counts["(other / unmatched)"] += 1
    return counts


def count_by_query(service) -> dict[str, int]:
    """ドメイン別件数は検索クエリで取る（metadata 全件取得より高速）。"""
    result: dict[str, int] = {}
    for domain in DOMAIN_KEYS:
        q = f"from:{domain}"
        ids = list_all_message_ids(service, q)
        result[domain] = len(ids)
    return result


def batch_trash(service, message_ids: list[str]) -> int:
    moved = 0
    for i in range(0, len(message_ids), BATCH_SIZE):
        chunk = message_ids[i : i + BATCH_SIZE]
        service.users().messages().batchModify(
            userId="me",
            body={
                "ids": chunk,
                "addLabelIds": ["TRASH"],
                "removeLabelIds": ["INBOX", "UNREAD"],
            },
        ).execute()
        moved += len(chunk)
        print(f"  ゴミ箱へ移動: {moved}/{len(message_ids)}")
    return moved


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(
        description="特定送信元メールを Gmail ゴミ箱へ一括移動（既定は dry-run）"
    )
    p.add_argument(
        "--exec",
        dest="do_exec",
        action="store_true",
        help="実際にゴミ箱へ移動する（省略時は dry-run）",
    )
    p.add_argument(
        "--query",
        default=DEFAULT_QUERY,
        help=f'Gmail 検索クエリ（既定: "{DEFAULT_QUERY}"）',
    )
    p.add_argument(
        "--detail",
        action="store_true",
        help="dry-run 時に From ヘッダから内訳を再集計（遅い）",
    )
    return p.parse_args()


def main() -> int:
    args = parse_args()
    mode = "EXEC" if args.do_exec else "DRY-RUN"
    print(f"=== 特定送信元メール → ゴミ箱移動 ({mode}) ===")
    print(f"クエリ: {args.query}")
    print()

    service = get_gmail_service()

    print("対象メッセージを収集中...")
    message_ids = list_all_message_ids(service, args.query)
    total = len(message_ids)
    print(f"合計: {total} 件")

    print("\nドメイン別内訳（各 from: クエリ）:")
    by_domain = count_by_query(service)
    for domain, n in by_domain.items():
        print(f"  {domain}: {n} 件")

    if args.detail and message_ids:
        print("\nFrom ヘッダからの内訳（詳細）:")
        detail = classify_by_domain(service, message_ids)
        for k, n in detail.most_common():
            print(f"  {k}: {n} 件")

    if not args.do_exec:
        print(
            "\ndry-run: ゴミ箱への移動は行っていません。"
            "\n実移動する場合は --exec を付けて再実行してください。"
        )
        return 0

    if total == 0:
        print("\n対象メールはありません。完了。")
        return 0

    print(f"\nゴミ箱へ移動を開始します（最大 {BATCH_SIZE} 件/バッチ）...")
    moved = batch_trash(service, message_ids)
    print(f"\n完了: {moved} 件をゴミ箱へ移動しました（完全削除はしていません）。")

    print("\n移動後の残件数確認:")
    remaining = list_all_message_ids(service, args.query)
    # ゴミ箱内は通常の検索に残る場合があるため、in:inbox を別途確認
    inbox_q = f"({args.query}) in:inbox"
    inbox_remaining = list_all_message_ids(service, inbox_q)
    print(f"  クエリ全体の残件（ゴミ箱含む可能性あり）: {len(remaining)} 件")
    print(f"  受信トレイ残件 ({inbox_q}): {len(inbox_remaining)} 件")
    if inbox_remaining:
        print("⚠ 受信トレイにまだ残っています。内容を確認してください。")
        return 1
    print("受信トレイに対象メールは残っていません。")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: trash-from-senders.mjs
#!/usr/bin/env node
/**
 * 指定送信元ドメインのメールを Gmail ゴミ箱へ一括移動する。
 *
 * 既定は dry-run（件数・ドメイン別内訳のみ表示）。
 * 実際の移動は --exec 指定時のみ。完全削除は行わない。
 *
 * 認証:
 *   credentials/credentials.json … OAuth クライアント
 *   credentials/token.json … 初回認可後に自動保存
 *
 * スコープ: https://www.googleapis.com/auth/gmail.modify
 *
 * 使い方:
 *   node trash-from-senders.mjs           # dry-run
 *   node trash-from-senders.mjs --exec    # 実際にゴミ箱へ移動
 */

import { createServer } from 'node:http';
import { readFileSync, writeFileSync, existsSync, mkdirSync } from 'node:fs';
import { createInterface } from 'node:readline';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { execFile } from 'node:child_process';
import { google } from 'googleapis';

const ROOT = dirname(fileURLToPath(import.meta.url));
const CRED_DIR = join(ROOT, 'credentials');
const CREDENTIALS_FILE = join(CRED_DIR, 'credentials.json');
const TOKEN_FILE = join(CRED_DIR, 'token.json');

const SCOPES = ['https://www.googleapis.com/auth/gmail.modify'];
const DEFAULT_QUERY = 'from:nikkeibp.co.jp OR from:media-radar.jp';
const DOMAIN_KEYS = ['nikkeibp.co.jp', 'media-radar.jp'];
const BATCH_SIZE = 1000;
const LIST_PAGE_SIZE = 500;
// credentials.json が web クライアントの場合は登録済み redirect を優先
function resolveRedirect(cfg) {
  const uris = cfg.redirect_uris || [];
  const preferred = uris.find((u) => /localhost|127\.0\.0\.1/.test(u));
  return preferred || 'http://127.0.0.1:53682';
}

function parseArgs(argv) {
  const args = { exec: false, query: DEFAULT_QUERY };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--exec') args.exec = true;
    else if (a === '--query' && argv[i + 1]) args.query = argv[++i];
  }
  return args;
}

async function authorize() {
  if (!existsSync(CREDENTIALS_FILE)) {
    console.error(
      `OAuth クライアントが見つかりません: ${CREDENTIALS_FILE}\n` +
        'Google Cloud Console で Desktop アプリの OAuth クライアント JSON を\n' +
        'credentials/credentials.json として配置してください。'
    );
    process.exit(2);
  }

  const raw = JSON.parse(readFileSync(CREDENTIALS_FILE, 'utf8'));
  const cfg = raw.installed || raw.web;
  if (!cfg) {
    console.error('credentials.json に installed / web クライアント情報がありません');
    process.exit(2);
  }

  const redirectUri = resolveRedirect(cfg);
  const oAuth2Client = new google.auth.OAuth2(
    cfg.client_id,
    cfg.client_secret,
    redirectUri
  );
  oAuth2Client.__redirectUri = redirectUri;

  if (existsSync(TOKEN_FILE)) {
    const tokens = JSON.parse(readFileSync(TOKEN_FILE, 'utf8'));
    oAuth2Client.setCredentials(tokens);
    const granted = new Set(tokens.scope ? tokens.scope.split(/\s+/) : []);
    const needScope = SCOPES.some((s) => !granted.has(s));
    if (!needScope) {
      try {
        const { credentials } = await oAuth2Client.refreshAccessToken();
        oAuth2Client.setCredentials(credentials);
        writeFileSync(TOKEN_FILE, JSON.stringify(credentials, null, 2));
        return oAuth2Client;
      } catch {
        // fall through
      }
    }
  }

  const authUrl = oAuth2Client.generateAuthUrl({
    access_type: 'offline',
    scope: SCOPES,
    prompt: 'select_account consent',
    login_hint: 'k.soeda@medi-canvas.com',
  });

  const code = await getAuthCode(authUrl, oAuth2Client.__redirectUri);
  const { tokens } = await oAuth2Client.getToken(code);
  oAuth2Client.setCredentials(tokens);
  mkdirSync(CRED_DIR, { recursive: true });
  writeFileSync(TOKEN_FILE, JSON.stringify(tokens, null, 2));
  console.log('認可完了。token.json を更新しました。');
  return oAuth2Client;
}

function getAuthCode(authUrl, redirectUri) {
  const listenUrl = new URL(redirectUri);
  const host = listenUrl.hostname === 'localhost' ? '127.0.0.1' : listenUrl.hostname;
  const port = Number(listenUrl.port || (listenUrl.protocol === 'https:' ? 443 : 80));

  return new Promise((resolve, reject) => {
    const server = createServer((req, res) => {
      try {
        const url = new URL(req.url, redirectUri);
        const code = url.searchParams.get('code');
        const err = url.searchParams.get('error');
        res.writeHead(200, { 'Content-Type': 'text/plain; charset=utf-8' });
        if (err) {
          res.end(`認可エラー: ${err}\nこのウィンドウを閉じてください。`);
          server.close();
          reject(new Error(err));
          return;
        }
        if (!code) {
          res.end('code がありません。');
          return;
        }
        res.end('認可成功。このウィンドウを閉じてターミナルに戻ってください。');
        server.close();
        resolve(code);
      } catch (e) {
        reject(e);
      }
    });

    server.listen(port, host, () => {
      console.log(`\n認可コールバック待機: ${redirectUri}`);
      console.log('ブラウザで Google 認可を開きます...');
      console.log(authUrl);
      const open =
        process.platform === 'darwin'
          ? 'open'
          : process.platform === 'win32'
            ? 'start'
            : 'xdg-open';
      execFile(open, [authUrl], () => {});
      console.log(
        '\nブラウザが開かない場合は上記 URL を開いてください。\n' +
          '（または、リダイレクト後の URL から code= を手動入力）\n'
      );
    });

    server.on('error', (e) => {
      console.error(
        `ローカルサーバー起動失敗 (${host}:${port}): ${e.message}\n` +
          '別プロセスがポートを使用しているか、redirect URI を確認してください。'
      );
      reject(e);
    });

    const rl = createInterface({ input: process.stdin, output: process.stdout });
    rl.question('認可コードを貼り付け（自動取得した場合は Enter のみ）: ', (line) => {
      rl.close();
      const trimmed = line.trim();
      if (trimmed) {
        server.close();
        resolve(trimmed);
      }
    });
  });
}

async function listAllMessageIds(gmail, query) {
  const ids = [];
  let pageToken;
  do {
    const resp = await gmail.users.messages.list({
      userId: 'me',
      q: query,
      pageToken,
      maxResults: LIST_PAGE_SIZE,
    });
    for (const m of resp.data.messages || []) {
      if (m.id) ids.push(m.id);
    }
    pageToken = resp.data.nextPageToken;
  } while (pageToken);
  return ids;
}

async function countByDomain(gmail) {
  const result = {};
  for (const domain of DOMAIN_KEYS) {
    const ids = await listAllMessageIds(gmail, `from:${domain}`);
    result[domain] = ids.length;
  }
  return result;
}

async function batchTrash(gmail, messageIds) {
  let moved = 0;
  for (let i = 0; i < messageIds.length; i += BATCH_SIZE) {
    const chunk = messageIds.slice(i, i + BATCH_SIZE);
    await gmail.users.messages.batchModify({
      userId: 'me',
      requestBody: {
        ids: chunk,
        addLabelIds: ['TRASH'],
        removeLabelIds: ['INBOX', 'UNREAD'],
      },
    });
    moved += chunk.length;
    console.log(`  ゴミ箱へ移動: ${moved}/${messageIds.length}`);
  }
  return moved;
}

async function main() {
  const args = parseArgs(process.argv);
  const mode = args.exec ? 'EXEC' : 'DRY-RUN';
  console.log(`=== 特定送信元メール → ゴミ箱移動 (${mode}) ===`);
  console.log(`クエリ: ${args.query}\n`);

  const auth = await authorize();
  const gmail = google.gmail({ version: 'v1', auth });

  const profile = await gmail.users.getProfile({ userId: 'me' });
  const email = profile.data.emailAddress || '(unknown)';
  console.log(`認可アカウント: ${email}`);
  // medi-canvas は send-as / エイリアスとして受信している想定。
  // ログイン本体は k.soeda.mediforce@gmail.com でも可。
  const allowed = new Set([
    'k.soeda@medi-canvas.com',
    'k.soeda.mediforce@gmail.com',
  ]);
  if (!allowed.has(email.toLowerCase())) {
    console.error(
      `\n⚠ 想定外のアカウントです（想定: medi-canvas / mediforce）。\n` +
        'credentials/token.json を削除して再実行し、正しいアカウントで認可してください。'
    );
    process.exit(3);
  }
  console.log();

  console.log('対象メッセージを収集中...');
  const messageIds = await listAllMessageIds(gmail, args.query);
  console.log(`合計: ${messageIds.length} 件`);

  console.log('\nドメイン別内訳（各 from: クエリ）:');
  const byDomain = await countByDomain(gmail);
  for (const [domain, n] of Object.entries(byDomain)) {
    console.log(`  ${domain}: ${n} 件`);
  }

  if (!args.exec) {
    console.log(
      '\ndry-run: ゴミ箱への移動は行っていません。\n' +
        '実移動する場合は --exec を付けて再実行してください。'
    );
    return;
  }

  if (messageIds.length === 0) {
    console.log('\n対象メールはありません。完了。');
    return;
  }

  console.log(
    `\nゴミ箱へ移動を開始します（最大 ${BATCH_SIZE} 件/バッチ）...`
  );
  const moved = await batchTrash(gmail, messageIds);
  console.log(
    `\n完了: ${moved} 件をゴミ箱へ移動しました（完全削除はしていません）。`
  );

  console.log('\n移動後の残件数確認:');
  const remaining = await listAllMessageIds(gmail, args.query);
  const inboxQ = `(${args.query}) in:inbox`;
  const inboxRemaining = await listAllMessageIds(gmail, inboxQ);
  console.log(`  クエリ全体の残件（ゴミ箱含む可能性あり）: ${remaining.length} 件`);
  console.log(`  受信トレイ残件 (${inboxQ}): ${inboxRemaining.length} 件`);
  if (inboxRemaining.length > 0) {
    console.error('⚠ 受信トレイにまだ残っています。内容を確認してください。');
    process.exit(1);
  }
  console.log('受信トレイに対象メールは残っていません。');
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
